### ISSUE TITLE
HashBuilder doesn't support embedded branch nodes

### ISSUE THREAD

**Issue by yperbasis** [ledgerwatch/erigon#1866]
`(*HashBuilder) branchHash` assumes that a branch node cannot be embedded (i.e. it assumes that branch nodes's RLP is no shorter than 32 bytes).

https://github.com/ledgerwatch/turbo-geth/blob/hash_builder_fix/turbo/trie/hashbuilder_test.go reproduces the issue.

**Comment by AlexeyAkhunov** [comment]
Wow, great, do you have a fix for this?

**Comment by yperbasis** [comment]
Nope, I don't have a fix yet. I fixed a somewhat similar issue in Silkworm, but haven't got to fixing it in TG yet. `HashBuilder's` code probably needs some love. If somebody wants to fix the issue, then by all means go. Otherwise I can work on it, but a bit later.

**Comment by mandrigin** [comment]
@yperbasis there is no more demo branch. If that issue still is relevant, please reopen.
